prism-agent/service/server/src/main/scala/io/iohk/atala/iam/authentication/oidc/KeycloakAuthenticatorImpl.scala

```diff
@@ -0,0 +1,103 @@
+package io.iohk.atala.iam.authentication.oidc
+
+import io.iohk.atala.agent.walletapi.service.WalletManagementService
+import io.iohk.atala.iam.authentication.AuthenticationError
+import io.iohk.atala.iam.authentication.AuthenticationError.AuthenticationMethodNotEnabled
+import io.iohk.atala.shared.models.WalletId
+import pdi.jwt.JwtCirce
+import pdi.jwt.JwtOptions
+import zio.*
+import zio.json.ast.Json
+
+import java.util.UUID
+import scala.util.Try
+
+class KeycloakAuthenticatorImpl(
+    client: KeycloakClient,
+    keycloakConfig: KeycloakConfig,
+    walletService: WalletManagementService
+) extends KeycloakAuthenticator {
+
+  override def isEnabled: Boolean = keycloakConfig.enabled
+
+  override def authenticate(token: String): IO[AuthenticationError, KeycloakEntity] = {
+    if (isEnabled) {
+      for {
+        introspection <- client.introspectToken(token)
+        _ <- ZIO
+          .fail(AuthenticationError.InvalidCredentials("The accessToken is invalid."))
+          .unless(introspection.active)
+        entityId <- ZIO
+          .fromOption(introspection.sub)
+          .mapError(_ => AuthenticationError.UnexpectedError("Subject ID is not found in the accessToken."))
+          .flatMap { id =>
+            ZIO
+              .attempt(UUID.fromString(id))
+              .mapError(e => AuthenticationError.UnexpectedError(s"Subject ID in accessToken is not a UUID. $e"))
+          }
+      } yield KeycloakEntity(entityId, token)
+    } else ZIO.fail(AuthenticationMethodNotEnabled("Keycloak authentication is not enabled"))
+  }
+
+  override def authorize(entity: KeycloakEntity): IO[AuthenticationError, WalletId] = {
+    val token = entity.rawToken
+    for {
+      isRpt <- inferIsRpt(entity.rawToken)
+      rptEffect =
+        if (isRpt) ZIO.succeed(token)
+        else if (keycloakConfig.autoUpgradeToRPT) client.getRpt(token)
+        else ZIO.fail(AuthenticationError.InvalidCredentials(s"AccessToken is not RPT."))
+      rpt <- rptEffect.logError("Fail to obtail RPT for wallet permissions")
+      permittedResources <- client.checkPermissions(rpt)
+      walletId <- getPermittedWallet(permittedResources)
+    } yield walletId
+  }
+
+  private def getPermittedWallet(resourceIds: Seq[String]): IO[AuthenticationError, WalletId] = {
+    val walletIds = resourceIds.flatMap(id => Try(UUID.fromString(id)).toOption).map(WalletId.fromUUID)
+    walletService
+      .getWallets(walletIds)
+      .mapError(e => AuthenticationError.UnexpectedError(e.toThrowable.getMessage()))
+      .flatMap {
+        case head +: Nil => ZIO.succeed(head.id)
+        case Nil =>
+          ZIO.fail(AuthenticationError.ResourceNotPermitted("No wallet permissions found."))
+        case ls =>
```

Comment:
probably, we can detect that the wallet access is ambiguous at the very beginning of this method, when `walletIds` length is >1

Reply:
This works if all the resources are wallet. As soon as there are other type of resources, we must check the validity of the IDs against the database.